File: recursion/lists/applied/add_two_numbers.scala
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */

// brute force approach
// go through each linked list to build the number - multiplying by 10 for each step
// then add the two numbers and then convert back into a linked list
// problem is O(3n)

// single pass
// loop through both lists in parallel - carrying over
// problem because of ordering we need to accumulate a listnode as well
// building in the wrong order - we could reverse but that would cost O(n)
// so can we build it the opposite way - maybe we have to accumulate a number and then output afterwards

object Solution {
    def addTwoNumbers(l1: ListNode, l2: ListNode): ListNode = {
        def recursiveAddTwoNumbers(l1: ListNode, l2: ListNode, carryOver : Int): ListNode = {
            (l1, l2) match {
                // base case
                case (null, null) if carryOver > 0 => ListNode(carryOver, null)
                case (null, null) => null
                // recursive case
                case (l1, null) => 
                    ListNode((l1.x + carryOver) % 10, 
                    recursiveAddTwoNumbers(l1.next, null,(l1.x + carryOver) / 10))
                case (null, l2) => 
                    ListNode((l2.x + carryOver) % 10, 
                    recursiveAddTwoNumbers(null, l2.next,(l2.x + carryOver) / 10))
                case (l1, l2) => 
                    ListNode((l1.x + l2.x + carryOver) % 10, 
                    recursiveAddTwoNumbers(l1.next, l2.next,(l1.x + l2.x + carryOver) / 10))
            }
        }
        recursiveAddTwoNumbers(l1, l2, 0)
    }
}

File: data_structures/constant_memory/rolling_sum/addTwoNumbers.go
package data_structures

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	// reverse-order = natural way to get number size is build from smallest digit
	// but would require either O(n) storage and O(2n) pass
	// single-pass = need to know power of 10 to multiply by so either 1. use len but requires O(n) or 2. O(1) storage and multiply on fly
	// building digits = single pass is intuitive, extract digits with / 10 remainder
	// choices: 1. recursive 2. for loop
	// choices: 1. calculate sum then create output 2. calculate sum and output at the same time
	cur := new(ListNode)
	ret := cur
	sum := 0
	for l1 != nil || l2 != nil {
		if l1 != nil {
			sum = sum + l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			sum = sum + l2.Val
			l2 = l2.Next
		}

		cur.Next = &ListNode{
			sum % 10,
			nil,
		}
		cur = cur.Next
		sum = sum / 10
	}

	if sum > 0 {
		cur.Next = &ListNode{sum, nil}
	}

	return ret.Next
}
